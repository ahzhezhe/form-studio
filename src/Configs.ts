import { ChoiceOnSelected, ChoiceValue, CustomConfigs, QuestionType } from './Types';

/**
 * @category Configs
 */
export interface ItemConfigs {
  /**
   * Item id.
   * It should be unique throughout the entire form.
   * It will be auto generated if you leave it falsy.
   */
  id?: string;
  /**
   * Item sort order.
   * Items will be sorted by their positions in the list if their order is `undefined`.
   */
  order?: number;
  /**
   * Whether or not this item is disabled by default.
   * The actual disabled property will be based on the logic that you define in other part of the configs.
   */
  defaultDisabled?: boolean;
  /**
   * Any values that help you determine on how to render the frontend UI for this item.
   */
  ui?: CustomConfigs;
}

/**
 * @category Configs
 */
export interface GroupConfigs extends ItemConfigs {
  /**
   * Sub-groups.
   */
  groups?: GroupConfigs[];
  /**
   * Questions under this group.
   */
  questions?: QuestionConfigs[];
}

/**
 * @category Configs
 */
export interface QuestionConfigs extends ItemConfigs {
  /**
   * Question type.
   *
   * Questions with `any` as [[type]] accept `any` as answer.
   * You shouldn't have to specify [[choices]] for this question.
   *
   * Questions with `single` as [[type]] accept [[ChoiceValue]] as answer.
   * You should specify [[choices]] for this question.
   * Usually radio button group will be used for this question.
   *
   * Questions with `multiple` as [[type]] accept [[ChoiceValue]][] as answer.
   * You should specify [[choices]] for this question.
   * Usually check box group will be used for this question.
   */
  type: QuestionType;
  /**
   * Choices for questions with `single` or `multiple` as [[type]].
   */
  choices?: ChoiceConfigs[];
  /**
   * Names of the validators to be used for validation when answer of this question is changed.
   *
   * Validators will be executed in sequence of their positions in the list.
   */
  validators?: string[];
  /**
   * Any values that help to validator to perform validation.
   */
  validation?: CustomConfigs;
  /**
   * Default answer.
   *
   * Questions with `any` as [[type]] accept `any` as answer.
   *
   * Questions with `single` as [[type]] accept [[ChoiceValue]] as answer.
   *
   * Questions with `multiple` as [[type]] accept [[ChoiceValue]][] as answer.
   */
  defaultAnswer?: any;
}

/**
 * @category Configs
 */
export interface ChoiceConfigs extends ItemConfigs {
  /**
   * Value of this choice. It will be used as answer of the question.
   * Id will be used as value if you leave it `undefined`.
   */
  value?: ChoiceValue;
  /**
   * Actions to be performed when this choice is toggled.
   */
  onSelected?: ChoiceOnSelected;
}

/**
 * Configs to construct a form.
 *
 * @category Configs
 */
export type Configs = GroupConfigs[];
