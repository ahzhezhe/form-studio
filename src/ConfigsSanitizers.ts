import { ChoiceConfigs, Configs, GroupConfigs, QuestionConfigs } from './Configs';
import { Choice, ExportedConfigs, Group, Question } from './ExportedConfigs';

export const sanitizeConfigs = (configs: Configs): ExportedConfigs => ({
  groups: configs.groups ? sanitizeGroupConfigs(undefined, configs.groups) : [],
  questions: configs.questions ? sanitizeQuestionConfigs(undefined, configs.questions) : []
});

const sanitizeGroupConfigs = (parentGroupId: string | undefined, groups: GroupConfigs[]): Group[] => groups.map((group, i) => {
  const id = group.id || (parentGroupId ? `${parentGroupId}_g${i}` : `g${i}`);
  return {
    id,
    defaultDisabled: !!group.defaultDisabled,
    custom: group.custom,
    groups: group.groups ? sanitizeGroupConfigs(id, group.groups) : [],
    questions: group.questions ? sanitizeQuestionConfigs(id, group.questions) : []
  };
});

const sanitizeQuestionConfigs = (groupId: string | undefined, questions: QuestionConfigs[]): Question[] => questions.map((question, i) => {
  const id = question.id || (groupId ? `${groupId}_q${i}` : `q${i}`);
  return {
    id,
    defaultDisabled: !!question.defaultDisabled,
    custom: question.custom,
    type: question.type,
    choices: question.type !== 'any' ? sanitizeChoiceConfigs(id, question.choices || []) : [],
    validators: question.validators || [],
    defaultAnswer: question.defaultAnswer
  };
});

const sanitizeChoiceConfigs = (questionId: string, choices: ChoiceConfigs[]): Choice[] => choices.map((choice, i) => {
  const id = choice.id || `${questionId}_c${i}`;
  return {
    id,
    defaultDisabled: !!choice.defaultDisabled,
    custom: choice.custom,
    value: choice.value === undefined || choice.value === null ? id : choice.value,
    onSelected: choice.onSelected || {}
  };
});
