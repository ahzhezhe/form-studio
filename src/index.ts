/* eslint-disable import/no-default-export */
import { GroupConfig, QuestionConfig, ChoiceConfig, Config } from './Configs';
import { FormEngine } from './FormEngine';
import { GroupInitConfig, QuestionInitConfig, ChoiceInitConfig, InitConfig } from './InitConfigs';
import { GroupTemplate, QuestionTemplate, ChoiceTemplate, Template } from './Templates';
import { CustomConfig, QuestionType, ChoiceValue, ChoiceOnSelected, Validator, Answers, Errors } from './Types';
import { useTemplate } from './useTemplate';

export {
  GroupConfig, QuestionConfig, ChoiceConfig, Config,
  GroupInitConfig, QuestionInitConfig, ChoiceInitConfig, InitConfig,
  GroupTemplate, QuestionTemplate, ChoiceTemplate, Template,
  CustomConfig, QuestionType, ChoiceValue, ChoiceOnSelected, Validator, Answers, Errors,
  useTemplate
};

export default FormEngine;
