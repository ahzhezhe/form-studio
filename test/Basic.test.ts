import { Form } from '../src/Form';
import { validators, getConfigs } from '.';

describe('Basic', () => {

  describe('No Default Answers', () => {

    const form = new Form(getConfigs(), validators);

    test('getConfigs', () => {
      expect(form.getConfigs()).toMatchSnapshot();
    });

    test('getRenderInstructions', () => {
      expect(form.getRenderInstructions()).toMatchSnapshot();
    });

    test('getErrors', () => {
      expect(form.getErrors()).toMatchSnapshot();
    });

    test('getCurrentAnswers', () => {
      expect(form.getCurrentAnswers()).toMatchSnapshot();
    });

    test('getValidatedAnswers', () => {
      expect(form.getValidatedAnswers()).toMatchSnapshot();
    });

  });

  describe('With Default Answers', () => {

    const form = new Form(getConfigs(true), validators);

    test('getConfigs', () => {
      expect(form.getConfigs()).toMatchSnapshot();
    });

    test('getRenderInstructions', () => {
      expect(form.getRenderInstructions()).toMatchSnapshot();
    });

    test('getErrors', () => {
      expect(form.getErrors()).toMatchSnapshot();
    });

    test('getCurrentAnswers', () => {
      expect(form.getCurrentAnswers()).toMatchSnapshot();
    });

    test('getValidatedAnswers', () => {
      expect(form.getValidatedAnswers()).toMatchSnapshot();
    });

  });

});
