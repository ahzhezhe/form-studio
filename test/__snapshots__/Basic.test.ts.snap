// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Basic No Default Answers getConfigs 1`] = `
Array [
  Object {
    "defaultDisabled": false,
    "groups": Array [
      Object {
        "defaultDisabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "defaultAnswer": undefined,
            "defaultDisabled": false,
            "id": "subGroup1Question1",
            "type": "any",
            "ui": Object {
              "title": "subGroup1Questions",
            },
            "validation": Object {},
            "validators": Array [
              "notNull",
            ],
          },
          Object {
            "choices": Array [
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question2Choice1",
                "onSelected": Object {
                  "enable": Array [
                    "subGroup1Question3",
                  ],
                },
                "ui": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question2Choice2",
                "onSelected": Object {
                  "disable": Array [
                    "subGroup1Question4",
                  ],
                },
                "ui": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question2Choice3",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "value": "subGroup1Question2Choice3",
              },
            ],
            "defaultAnswer": undefined,
            "defaultDisabled": false,
            "id": "subGroup1Question2",
            "type": "single",
            "ui": Object {
              "title": "subGroup1Question2",
            },
            "validation": Object {},
            "validators": Array [
              "notNullSingle",
            ],
          },
          Object {
            "choices": Array [
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question3Choice1",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question3Choice2",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question3Choice3",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "value": "subGroup1Question3Choice3",
              },
            ],
            "defaultAnswer": undefined,
            "defaultDisabled": true,
            "id": "subGroup1Question3",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validation": Object {},
            "validators": Array [
              "atLeast1",
            ],
          },
          Object {
            "choices": Array [
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question4Choice1",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question4Choice2",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question4Choice3",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "value": "subGroup1Question4Choice3",
              },
            ],
            "defaultAnswer": undefined,
            "defaultDisabled": false,
            "id": "subGroup1Question4",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validation": Object {},
            "validators": Array [
              "atLeast1",
            ],
          },
        ],
        "ui": Object {},
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "defaultAnswer": undefined,
        "defaultDisabled": false,
        "id": "group1Question1",
        "type": "any",
        "ui": Object {
          "title": "group1Questions",
        },
        "validation": Object {},
        "validators": Array [
          "notNull",
        ],
      },
      Object {
        "choices": Array [
          Object {
            "defaultDisabled": false,
            "id": "group1Question2Choice1",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question2Choice1",
            },
            "value": "group1Question2Choice1",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question2Choice2",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question2Choice2",
            },
            "value": "group1Question2Choice2",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question2Choice3",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question2Choice3",
            },
            "value": "group1Question2Choice3",
          },
        ],
        "defaultAnswer": undefined,
        "defaultDisabled": false,
        "id": "group1Question2",
        "type": "single",
        "ui": Object {
          "title": "group1Question2",
        },
        "validation": Object {},
        "validators": Array [
          "notNullSingle",
        ],
      },
      Object {
        "choices": Array [
          Object {
            "defaultDisabled": false,
            "id": "group1Question3Choice1",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question3Choice1",
            },
            "value": "group1Question3Choice1",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question3Choice2",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question3Choice2",
            },
            "value": "group1Question3Choice2",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question3Choice3",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question3Choice3",
            },
            "value": "group1Question3Choice3",
          },
        ],
        "defaultAnswer": undefined,
        "defaultDisabled": false,
        "id": "group1Question3",
        "type": "multiple",
        "ui": Object {
          "title": "group1Question3",
        },
        "validation": Object {},
        "validators": Array [],
      },
    ],
    "ui": Object {},
  },
]
`;

exports[`Basic No Default Answers getCurrentAnswers 1`] = `
Object {
  "group1Question3": Array [],
  "subGroup1Question3": Array [],
  "subGroup1Question4": Array [],
}
`;

exports[`Basic No Default Answers getErrors 1`] = `
Object {
  "group1Question1": "This field is required.",
  "group1Question2": "Please select an option.",
  "subGroup1Question1": "This field is required.",
  "subGroup1Question2": "Please select an option.",
  "subGroup1Question4": "Please select at least 1 option.",
}
`;

exports[`Basic No Default Answers getRenderInstructions 1`] = `
Array [
  Object {
    "disabled": false,
    "groups": Array [
      Object {
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": undefined,
            "disabled": false,
            "error": "This field is required.",
            "id": "subGroup1Question1",
            "type": "any",
            "ui": Object {
              "title": "subGroup1Questions",
            },
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "ui": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "ui": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "ui": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": undefined,
            "disabled": false,
            "error": "Please select an option.",
            "id": "subGroup1Question2",
            "type": "single",
            "ui": Object {
              "title": "subGroup1Question2",
            },
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": true,
                "id": "subGroup1Question3Choice1",
                "ui": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "disabled": true,
                "id": "subGroup1Question3Choice2",
                "ui": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "disabled": true,
                "id": "subGroup1Question3Choice3",
                "ui": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [],
            "disabled": true,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question3",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": undefined,
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "ui": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "ui": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "ui": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [],
            "disabled": false,
            "error": "Please select at least 1 option.",
            "id": "subGroup1Question4",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": undefined,
            "validating": false,
          },
        ],
        "ui": Object {},
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": undefined,
        "disabled": false,
        "error": "This field is required.",
        "id": "group1Question1",
        "type": "any",
        "ui": Object {
          "title": "group1Questions",
        },
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question2Choice1",
            "ui": Object {
              "title": "group1Question2Choice1",
            },
            "value": "group1Question2Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice2",
            "ui": Object {
              "title": "group1Question2Choice2",
            },
            "value": "group1Question2Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice3",
            "ui": Object {
              "title": "group1Question2Choice3",
            },
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": undefined,
        "disabled": false,
        "error": "Please select an option.",
        "id": "group1Question2",
        "type": "single",
        "ui": Object {
          "title": "group1Question2",
        },
        "validatedAnswer": undefined,
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question3Choice1",
            "ui": Object {
              "title": "group1Question3Choice1",
            },
            "value": "group1Question3Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice2",
            "ui": Object {
              "title": "group1Question3Choice2",
            },
            "value": "group1Question3Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice3",
            "ui": Object {
              "title": "group1Question3Choice3",
            },
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [],
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "multiple",
        "ui": Object {
          "title": "group1Question3",
        },
        "validatedAnswer": Array [],
        "validating": false,
      },
    ],
    "ui": Object {},
  },
]
`;

exports[`Basic No Default Answers getValidatedAnswers 1`] = `
Object {
  "group1Question3": Array [],
}
`;

exports[`Basic With Default Answers getConfigs 1`] = `
Array [
  Object {
    "defaultDisabled": false,
    "groups": Array [
      Object {
        "defaultDisabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "defaultAnswer": "subGroup1Question1Answer",
            "defaultDisabled": false,
            "id": "subGroup1Question1",
            "type": "any",
            "ui": Object {
              "title": "subGroup1Questions",
            },
            "validation": Object {},
            "validators": Array [
              "notNull",
            ],
          },
          Object {
            "choices": Array [
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question2Choice1",
                "onSelected": Object {
                  "enable": Array [
                    "subGroup1Question3",
                  ],
                },
                "ui": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question2Choice2",
                "onSelected": Object {
                  "disable": Array [
                    "subGroup1Question4",
                  ],
                },
                "ui": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question2Choice3",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "value": "subGroup1Question2Choice3",
              },
            ],
            "defaultAnswer": "subGroup1Question2Choice1",
            "defaultDisabled": false,
            "id": "subGroup1Question2",
            "type": "single",
            "ui": Object {
              "title": "subGroup1Question2",
            },
            "validation": Object {},
            "validators": Array [
              "notNullSingle",
            ],
          },
          Object {
            "choices": Array [
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question3Choice1",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question3Choice2",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question3Choice3",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "value": "subGroup1Question3Choice3",
              },
            ],
            "defaultAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "defaultDisabled": true,
            "id": "subGroup1Question3",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validation": Object {},
            "validators": Array [
              "atLeast1",
            ],
          },
          Object {
            "choices": Array [
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question4Choice1",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question4Choice2",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "defaultDisabled": false,
                "id": "subGroup1Question4Choice3",
                "onSelected": Object {},
                "ui": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "value": "subGroup1Question4Choice3",
              },
            ],
            "defaultAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "defaultDisabled": false,
            "id": "subGroup1Question4",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validation": Object {},
            "validators": Array [
              "atLeast1",
            ],
          },
        ],
        "ui": Object {},
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "defaultAnswer": "group1Question1Answer",
        "defaultDisabled": false,
        "id": "group1Question1",
        "type": "any",
        "ui": Object {
          "title": "group1Questions",
        },
        "validation": Object {},
        "validators": Array [
          "notNull",
        ],
      },
      Object {
        "choices": Array [
          Object {
            "defaultDisabled": false,
            "id": "group1Question2Choice1",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question2Choice1",
            },
            "value": "group1Question2Choice1",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question2Choice2",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question2Choice2",
            },
            "value": "group1Question2Choice2",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question2Choice3",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question2Choice3",
            },
            "value": "group1Question2Choice3",
          },
        ],
        "defaultAnswer": "group1Question2Choice1",
        "defaultDisabled": false,
        "id": "group1Question2",
        "type": "single",
        "ui": Object {
          "title": "group1Question2",
        },
        "validation": Object {},
        "validators": Array [
          "notNullSingle",
        ],
      },
      Object {
        "choices": Array [
          Object {
            "defaultDisabled": false,
            "id": "group1Question3Choice1",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question3Choice1",
            },
            "value": "group1Question3Choice1",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question3Choice2",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question3Choice2",
            },
            "value": "group1Question3Choice2",
          },
          Object {
            "defaultDisabled": false,
            "id": "group1Question3Choice3",
            "onSelected": Object {},
            "ui": Object {
              "title": "group1Question3Choice3",
            },
            "value": "group1Question3Choice3",
          },
        ],
        "defaultAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "defaultDisabled": false,
        "id": "group1Question3",
        "type": "multiple",
        "ui": Object {
          "title": "group1Question3",
        },
        "validation": Object {},
        "validators": Array [],
      },
    ],
    "ui": Object {},
  },
]
`;

exports[`Basic With Default Answers getCurrentAnswers 1`] = `
Object {
  "group1Question1": "group1Question1Answer",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Answer",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;

exports[`Basic With Default Answers getErrors 1`] = `Object {}`;

exports[`Basic With Default Answers getRenderInstructions 1`] = `
Array [
  Object {
    "disabled": false,
    "groups": Array [
      Object {
        "disabled": false,
        "groups": Array [],
        "id": "subGroup1",
        "questions": Array [
          Object {
            "choices": Array [],
            "currentAnswer": "subGroup1Question1Answer",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question1",
            "type": "any",
            "ui": Object {
              "title": "subGroup1Questions",
            },
            "validatedAnswer": "subGroup1Question1Answer",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice1",
                "ui": Object {
                  "title": "subGroup1Question2Choice1",
                },
                "value": "subGroup1Question2Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice2",
                "ui": Object {
                  "title": "subGroup1Question2Choice2",
                },
                "value": "subGroup1Question2Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question2Choice3",
                "ui": Object {
                  "title": "subGroup1Question2Choice3",
                },
                "value": "subGroup1Question2Choice3",
              },
            ],
            "currentAnswer": "subGroup1Question2Choice1",
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question2",
            "type": "single",
            "ui": Object {
              "title": "subGroup1Question2",
            },
            "validatedAnswer": "subGroup1Question2Choice1",
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice1",
                "ui": Object {
                  "title": "subGroup1Question3Choice1",
                },
                "value": "subGroup1Question3Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice2",
                "ui": Object {
                  "title": "subGroup1Question3Choice2",
                },
                "value": "subGroup1Question3Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question3Choice3",
                "ui": Object {
                  "title": "subGroup1Question3Choice3",
                },
                "value": "subGroup1Question3Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question3",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question3Choice1",
              "subGroup1Question3Choice2",
            ],
            "validating": false,
          },
          Object {
            "choices": Array [
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice1",
                "ui": Object {
                  "title": "subGroup1Question4Choice1",
                },
                "value": "subGroup1Question4Choice1",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice2",
                "ui": Object {
                  "title": "subGroup1Question4Choice2",
                },
                "value": "subGroup1Question4Choice2",
              },
              Object {
                "disabled": false,
                "id": "subGroup1Question4Choice3",
                "ui": Object {
                  "title": "subGroup1Question4Choice3",
                },
                "value": "subGroup1Question4Choice3",
              },
            ],
            "currentAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "disabled": false,
            "error": undefined,
            "id": "subGroup1Question4",
            "type": "multiple",
            "ui": Object {
              "title": "subGroup1Question3",
            },
            "validatedAnswer": Array [
              "subGroup1Question4Choice1",
              "subGroup1Question4Choice2",
            ],
            "validating": false,
          },
        ],
        "ui": Object {},
      },
    ],
    "id": "group1",
    "questions": Array [
      Object {
        "choices": Array [],
        "currentAnswer": "group1Question1Answer",
        "disabled": false,
        "error": undefined,
        "id": "group1Question1",
        "type": "any",
        "ui": Object {
          "title": "group1Questions",
        },
        "validatedAnswer": "group1Question1Answer",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question2Choice1",
            "ui": Object {
              "title": "group1Question2Choice1",
            },
            "value": "group1Question2Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice2",
            "ui": Object {
              "title": "group1Question2Choice2",
            },
            "value": "group1Question2Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question2Choice3",
            "ui": Object {
              "title": "group1Question2Choice3",
            },
            "value": "group1Question2Choice3",
          },
        ],
        "currentAnswer": "group1Question2Choice1",
        "disabled": false,
        "error": undefined,
        "id": "group1Question2",
        "type": "single",
        "ui": Object {
          "title": "group1Question2",
        },
        "validatedAnswer": "group1Question2Choice1",
        "validating": false,
      },
      Object {
        "choices": Array [
          Object {
            "disabled": false,
            "id": "group1Question3Choice1",
            "ui": Object {
              "title": "group1Question3Choice1",
            },
            "value": "group1Question3Choice1",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice2",
            "ui": Object {
              "title": "group1Question3Choice2",
            },
            "value": "group1Question3Choice2",
          },
          Object {
            "disabled": false,
            "id": "group1Question3Choice3",
            "ui": Object {
              "title": "group1Question3Choice3",
            },
            "value": "group1Question3Choice3",
          },
        ],
        "currentAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "disabled": false,
        "error": undefined,
        "id": "group1Question3",
        "type": "multiple",
        "ui": Object {
          "title": "group1Question3",
        },
        "validatedAnswer": Array [
          "group1Question3Choice1",
          "group1Question3Choice2",
        ],
        "validating": false,
      },
    ],
    "ui": Object {},
  },
]
`;

exports[`Basic With Default Answers getValidatedAnswers 1`] = `
Object {
  "group1Question1": "group1Question1Answer",
  "group1Question2": "group1Question2Choice1",
  "group1Question3": Array [
    "group1Question3Choice1",
    "group1Question3Choice2",
  ],
  "subGroup1Question1": "subGroup1Question1Answer",
  "subGroup1Question2": "subGroup1Question2Choice1",
  "subGroup1Question3": Array [
    "subGroup1Question3Choice1",
    "subGroup1Question3Choice2",
  ],
  "subGroup1Question4": Array [
    "subGroup1Question4Choice1",
    "subGroup1Question4Choice2",
  ],
}
`;
